File: build_all_data/land_cover/code/build_land_cover_conus.R
#building a land cover database
if (!require(terra)) install.packages('terra')
if (!require(sf)) install.packages('sf')
if (!require(tidyverse)) install.packages('tidyverse')
if (!require(readxl)) install.packages('readxl')
library(terra)
library(sf)
library(tidyverse)
library(readxl)

rm(list=ls())

#upload data on incidents
incidents <- st_read("/home/dashaa/wildfire_politics/data/build_wildfire_reports/output/incidents_spatial.gpkg")
incidents <- filter(incidents,POO_state!="AK")
incidents <- subset(incidents,select=c(incident_id,start_year,final_acres))
incidents <- st_transform(incidents, crs = st_crs("+proj=utm +zone=12 +datum=WGS84"))
incidents$final_sq_m <- incidents$final_acres*4046.8564224
incidents$radius <- sqrt(incidents$final_sq_m/pi)
incidents <- st_buffer(incidents,incidents$radius)
incidents <- st_transform(incidents, 
                          crs = "+proj=aea +lat_0=23 +lon_0=-96 +lat_1=29.5 +lat_2=45.5 +x_0=0 +y_0=0 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs ")

incidents$tree_cover_p <- 0
incidents$shrub_cover_p <- 0
incidents$herb_cover_p <- 0

path <- "/home/dashaa/wildfire_politics/data/build_land_cover"
input_path <- paste0(path,"/input/raw/cms_conus")
output_path <- paste0(path,"/output")

#insert a file with all links, years, types of cover - excel file
wget_commands <- read_excel(paste0(input_path,"/wget_commands3.xlsx"))
wget_commands$command <- paste0('wget -r -np -nH --reject "index.html*" -e robots=off ',wget_commands$link)

for (i in 1:nrow(wget_commands)) {
  
  year <- wget_commands$year[i]
  type <- wget_commands$type[i]
  incidents_year <- filter(incidents, start_year==year)
  
  setwd(input_path)
  system(wget_commands$command[i])
  
  files <- list.files(input_path,recursive=T,pattern=".tif",full.names = T)
  land_rasts <- lapply(files, rast, lyrs=1) #get the first layer from each file - mean estimate numbers
  
  for (land_rast in land_rasts) {
    
    #tc_rast[tc_rast==65535] <- NA
    
    land_values <- terra::extract(land_rast,incidents_year,mean,na.rm=TRUE)
    land_values[is.na(land_values)] <- 0
    
    if (type=="TC") {
      incidents_year$tree_cover_p <- incidents_year$tree_cover_p + rowMeans(land_values[2])
    } else if (type=="SC") {
      incidents_year$shrub_cover_p <- incidents_year$shrub_cover_p + rowMeans(land_values[2])
    } else {
      incidents_year$herb_cover_p <- incidents_year$herb_cover_p + rowMeans(land_values[2])
    }

  }
  
  incidents_year <- st_set_geometry(incidents_year,NULL)
  
  if (type=="TC") {
    incidents_year <- subset(incidents_year,select=c(incident_id,tree_cover_p))
    write.csv(incidents_year,paste0(path,"/input/built/conus/tree_cover/",year,".csv"))
  } else if (type=="SC") {
    incidents_year <- subset(incidents_year,select=c(incident_id,shrub_cover_p))
    write.csv(incidents_year,paste0(path,"/input/built/conus/shrub_cover/",year,".csv"))
  } else {
    incidents_year <- subset(incidents_year,select=c(incident_id,herb_cover_p))
    write.csv(incidents_year,paste0(path,"/input/built/conus/herb_cover/",year,".csv"))
  }
  
  unlink(files)
  
}

File: build_all_data/land_cover/code/build_land_cover_conus.py
#building land cover variables for contiguous U.S.
import os
import geopandas as gpd
import rasterio
import numpy as np
import pandas as pd
from shapely.geometry import Point, Polygon
from rasterio.mask import mask
from pyproj import Transformer

# Define paths
path = "/home/dashaa/wildfire_politics/data/build_land_cover"
input_path = os.path.join(path, "input/raw/above/Annual_Landcover_ABoVE_1691/data")
output_path = os.path.join(path, "output")

# Load data on incidents
incidents = gpd.read_file("/home/dashaa/wildfire_politics/data/build_wildfire_reports/output/incidents_spatial.gpkg")

# Filter for 1999 contiguous US incidents in Alaska
incidents = incidents[incidents['POO_state'] == "AK"]

# Subset and transform
incidents = incidents[['incident_id', 'start_year', 'final_acres']]
incidents = incidents.to_crs(epsg=32612)  # UTM zone 12N
incidents['final_sq_m'] = incidents['final_acres'] * 4046.8564224
incidents['radius'] = np.sqrt(incidents['final_sq_m'] / np.pi)
incidents['geometry'] = incidents.geometry.buffer(incidents['radius'])

# Transform to Albers Equal Area
incidents = incidents.to_crs("+proj=aea +lat_0=40 +lon_0=-96 +lat_1=50 +lat_2=70 +x_0=0 +y_0=0 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs")

# Initialize cover percentage columns
incidents['tree_cover_p'] = 0
incidents['shrub_cover_p'] = 0
incidents['herb_cover_p'] = 0

# List .tif files
files = [os.path.join(input_path, f) for f in os.listdir(input_path) if f.endswith('.tif')]

def count_cells_equal_to_value(raster, geom, values):
    with rasterio.open(raster) as src:
        out_image, out_transform = mask(src, [geom], crop=True)
        out_image = out_image[0]  # Take the first band
        cell_count = np.sum(np.isin(out_image, values))
        total_cells = out_image.size
        value_proportion = cell_count / total_cells
    return value_proportion

# Loop through years
new_incidents = None

for year in range(1998, 2013 + 1):
    rasts_year = [(raster, year - 1983) for raster in files]
    incidents_year = incidents[incidents['start_year'] == year + 1].copy()
    
    for raster, layer in rasts_year:
        tree_values = np.array([count_cells_equal_to_value(raster, geom, [1, 2]) for geom in incidents_year.geometry])
        shrub_values = np.array([count_cells_equal_to_value(raster, geom, [3]) for geom in incidents_year.geometry])
        herb_values = np.array([count_cells_equal_to_value(raster, geom, [4]) for geom in incidents_year.geometry])
        
        incidents_year['tree_cover_p'] += tree_values
        incidents_year['shrub_cover_p'] += shrub_values
        incidents_year['herb_cover_p'] += herb_values
    
    if new_incidents is None:
        new_incidents = incidents_year
    else:
        new_incidents = pd.concat([new_incidents, incidents_year], ignore_index=True)

# Convert cover percentages to percentages
new_incidents['tree_cover_p'] *= 100
new_incidents['shrub_cover_p'] *= 100
new_incidents['herb_cover_p'] *= 100

# Drop geometry and select necessary columns
new_incidents = new_incidents.drop(columns='geometry')
new_incidents = new_incidents[['incident_id', 'tree_cover_p', 'shrub_cover_p', 'herb_cover_p']]

# Save to CSV
new_incidents.to_csv(os.path.join(path, "input/built/alaska/land_cover.csv"), index=False)

